ethcore/sync/src/chain/mod.rs

```diff
@@ -700,6 +700,7 @@ impl ChainSync {
 	fn complete_sync(&mut self, io: &mut SyncIo) {
 		trace!(target: "sync", "Sync complete");
 		self.reset(io);
+		self.state = SyncState::Idle;
```

Comment:
Reset is setting the state to `ChainSync::get_init_state(self.warp_sync, io.chain());` already. Setting it to `Idle` here will break `warp-barrier`.

Reply:
Oups, you're right @tomusdrw !

Reply:
You're right it does break the warp barrier, but I also don't think the current behavior of setting to `get_init_state` is correct, that's why I forced it to `Idle`. Every time we're synced with the tip of the chain we go into `WaitingPeers` state which tries to trigger a snapshot sync. If you monitor `eth_syncing` it goes into syncing state briefly every time a new block is imported.

Reply:
Does it make sense to set it to `get_init_state` until `WAIT_PEERS_TIMEOUT` is elapsed, and force it to idle afterwards?

Reply:
I think this line: https://github.com/paritytech/parity-ethereum/blob/master/ethcore/sync/src/chain/mod.rs#L451 should be modified as you are saying, set to `Idle` when `WAIT_PEERS_TIMEOUT` is elapsed

Reply:
Okay I was wrong. This doesn't break the warp-barrier, it keeps on searching for the given snapshot until it finds it. It doesn't break the warp barrier because it will be stuck in `WaitingPeers` state and therefore will never complete any sync round (i.e. `complete_sync` won't be called which would force the state to `Idle`). It does break the informant because of the changes I made to `is_major_importing`, it will print as if the node was syncing while it is actually searching for the snapshot. Though I'm not sure what is the correct semantics here? When `eth_syncing` returns `false` should it mean that the node just isn't syncing, or that it's not syncing because it doesn't need to? I.e. it is already synced to the tip of the chain.

Reply:
Yeah, I guess if we're in `complete_sync`, then we've already reached warp barrier. We can make `reset` take an additional state param and pass either `ChainSync::get_init_state(self.warp_sync, io.chain())` or `SyncState::Idle`, but I don't see it as blocking this PR.

Reply:
I think the fix should be moved from `complete_sync()` to `reset()` – it's brittle to have the state first set in `reset()` and then clobbered right afterwards in `complete_sync()`. 
I am new to this code and might be completely wrong here, but it seems like we're conflating state with configuration (`WarpSync::Enabled`). Maybe I'm making the wrong assumption here, but isn't it true that once warp syncing is done we want it turned off until the node is restarted? Or are there cases when a node reaches the tip of the chain at one point t and starts syncing normally and then at t + n need to go back to warp syncing? If not, then `Idle` would be the correct state as soon as `complete_sync` is called?
In other words: if we're beyond the warp barrier or at the tip, calling `complete_sync()` should put us in `Idle` (but I'd move that logic to `reset()` if possible).

Reply:
> Or are there cases when a node reaches the tip of the chain at one point t and starts syncing normally and then at t + n need to go back to warp syncing?

Yes, we can switch back to warp syncing e.g. if our node went offline for a while (or just slow) and is too far away (`SNAPSHOT_RESTORE_THRESHOLD`) from the current tip (see [`maybe_start_snapshot_sync`](https://github.com/paritytech/parity-ethereum/blob/6d06c7ad183a363c8c470278fc5f37704af5f789/ethcore/sync/src/chain/mod.rs#L548-L613), which is called regularly on timer).